### DISCUSSION TITLE
[Guide] MiniMax H3 Ref2VA-Omni LoRA on 20GB/48GB GPUs: Windows setup, cache verification, and block-swap workflow

### DISCUSSION THREAD

**Discussion by lcdrzxlcdrzx** [kohya-ss/musubi-tuner#1061] [General]
# MiniMax H3 Ref2VA-Omni LoRA on 20GB/48GB GPUs: Windows setup, cache verification, and block-swap workflow

This is a practical, hardware-constrained setup report for MiniMax H3 Ref2VA-Omni LoRA training with musubi-tuner. It is intended for users who want a reproducible path from a clean Windows or Linux machine to a verified 1-step run before committing to a long GPU rental.

This guide intentionally contains no dataset, captions, model outputs, server endpoints, or account details.

## Tested scope

- MiniMax H3 Ref2VA-Omni LoRA
- Ref2VA pruned INT8 ConvRot transformer
- Qwen3-VL NVFP4 AWQ text encoder
- Video VAE + Audio VAE
- BF16, AdamW8bit, LoRA rank 16, batch size 1
- 73-frame videos (the H3 `17k+5` frame grid)
- Windows 10 + RTX 3080 20GB for a 1-step smoke test
- Linux + RTX 3090 48GB for 768x1344 full training preparation

## The reliable sequence

Do not begin with a long run. The following order isolated every real problem cheaply:

1. Create an isolated Python environment and verify CUDA + BF16.
2. Verify every model by exact byte count, SHA-256, and safetensors readability.
3. Validate the dataset JSONL: every target file exists and every caption is non-empty.
4. Build latent cache, then text cache, and count both cache types.
5. Run a 1-step LoRA smoke test.
6. Record VRAM peak and actual step time.
7. Only then choose resolution, block swap level, checkpoint interval, and max steps.

The 1-step test must complete forward, backward, optimizer update, LoRA save, and accelerator state save. A progress bar alone is not enough.

## Model file verification

An interrupted aria2 download can be shown as `OK` while the model remains unusable. An old `.aria2` control file can incorrectly treat a truncated file as complete, and sparse/preallocated files can appear to have the expected file size.

For every model file, use all three checks:

```bash
# 1. exact expected byte count
ls -l /path/to/model.safetensors

# 2. official SHA-256
sha256sum /path/to/model.safetensors

# 3. actual safetensors readability
python -c "from safetensors import safe_open; f=safe_open('/path/to/model.safetensors', framework='pt'); print(len(f.keys()))"
```

For large downloads, a workable aria2 pattern is:

```text
--continue=true --max-connection-per-server=16 --split=16 --min-split-size=20M --file-allocation=none
```

If byte count or SHA does not match, remove the stale control metadata and re-download rather than trusting the terminal summary.

## Windows isolation: clear PYTHONPATH

On a desktop environment with another Python application installed, inherited `PYTHONPATH` can load incompatible packages into the training environment.

For Windows `.bat` launchers:

```bat
set "PYTHONPATH="
```

For a shell command:

```bash
env -u PYTHONPATH python ...
```

Verify imports from the intended virtual environment before caching or training.

## 20GB 1-step smoke test profile

The following was sufficient to prove the full Ref2VA-Omni training path on a 20GB GPU. It is a smoke-test profile, not a throughput recommendation:

```text
--int8_convrot_base
--h3_training_mode ref2va_omni
--network_module networks.lora_minimax_h3
--network_dim 16
--network_alpha 16
--sdpa
--mixed_precision bf16
--gradient_checkpointing
--blocks_to_swap 50
--block_swap_h2d_only
--block_swap_ring_size 1
--block_swap_granularity layer
--optimizer_type AdamW8bit
--learning_rate 1e-4
--save_state
--save_state_on_train_end
--max_train_steps 1
```

This made the 1-step path fit on 20GB. It should not be copied unchanged to a larger GPU: swapping 50 layers is intentionally conservative and severely harms throughput.

## 48GB starting point

For 768x1344, 73 frames, batch size 1, start more conservatively than a no-swap assumption and measure:

```text
--gradient_checkpointing
--blocks_to_swap 24
--block_swap_h2d_only
--block_swap_ring_size 2
```

If it OOMs, increase swap progressively, e.g. 32 then 40. If it fits, reduce swap to improve throughput. On our 48GB setup, a 40-layer swap configuration ran stably but was slow; full training was roughly 123 seconds per optimizer step. This is a real measured result for this configuration, not a general speed claim.

Use a 1-step run after each significant change to resolution, frames, model, swap level, reference mode, or LoRA rank.

## Dataset cache integrity

Before training, verify:

- latent cache count matches target sample count;
- text cache count matches target sample count;
- every cache pair is readable;
- cache shape/resolution match the current TOML.

The trainer does not rebuild missing caches during a normal training run.

## Train/validation split: use independent cache directories

Do not point training and validation TOML files at one shared cache directory. The H3 bucket builder can enumerate cache files from the directory, so samples outside the JSONL split may be picked up.

Use separate directories, for example:

```text
cache/train_768/  # only training latent + text pairs
cache/val_768/    # only validation latent + text pairs
```

Then verify the log reports exactly the intended training count and batch count.

Full reproduction and proposed upstream fix: #1060.

## INT8 backward compatibility

With the released pruned INT8 ConvRot Ref2VA checkpoint, the default INT8 backward route can fail with a cuBLAS unsupported-status error on a non-16-aligned pruned dimension. We used the existing transient/BF16 route as a temporary local workaround and verified it through a full 1-step LoRA run.

Details and a proposed upstream fix: #1059.

Before applying local source edits, check whether the upstream issue has been fixed in your musubi-tuner revision.

## Long runs on interruptible rentals

Run training inside tmux or another session manager. Save full Accelerate state frequently, not only LoRA weights:

```text
--save_every_n_steps 20
--save_state
--autoresume
```

A complete state should include model weights, optimizer, scheduler, dataloader sampler, and random state. Test a resume from a saved state before relying on it for a long run.

## Monitoring

Use TensorBoard for:

- `loss/current`: noisy per-batch signal;
- `loss/average`: local cumulative average; it resets when a resumed process creates a new run;
- `val/loss`, `val/loss/video`, `val/loss/audio`: most useful for overfitting decisions;
- `grad/norm`: detect exploding/invalid gradients.

A single validation point is not a trend. For a small split, compare multiple checkpoints with the same validation set and fixed generation prompts/seeds.

## Practical stopping criteria

There is no universal absolute flow-matching loss target. The useful decision signal is:

- train loss down + validation loss down/stable: continue;
- train loss down + validation loss rising across checkpoints: stop or select an earlier checkpoint;
- NaN/Inf, abrupt persistent gradient growth, or validation failure: investigate before spending more GPU time.

For small video datasets, generate fixed-prompt comparison samples at multiple checkpoints rather than assuming the final step is best.

## Summary

The important lesson is not that a 20GB GPU is fast. It is that it can be used to validate the complete pipeline before a more expensive rental. Model integrity checks, cached-data accounting, 1-step backward validation, isolated validation caches, and frequent resume states prevent most costly failures.

---

# 中文版：MiniMax H3 Ref2VA-Omni LoRA 在 20GB / 48GB GPU 上的部署、缓存验收与 Block Swap 实战

这是一次受显存限制的 MiniMax H3 Ref2VA-Omni LoRA 实战配置报告，面向希望在干净 Windows 或 Linux 环境中，先完成可靠的 1-step 验证，再投入长时间 GPU 租赁的用户。

本文刻意不包含数据集、caption、模型输出、服务器地址、端口或账号信息。

## 已验证范围

- MiniMax H3 Ref2VA-Omni LoRA
- Ref2VA pruned INT8 ConvRot transformer
- Qwen3-VL NVFP4 AWQ text encoder
- Video VAE + Audio VAE
- BF16、AdamW8bit、LoRA rank 16、batch size 1
- 73 帧视频（符合 H3 的 `17k+5` 帧格）
- Windows 10 + RTX 3080 20GB：用于 1-step 冒烟测试
- Linux + RTX 3090 48GB：用于 768x1344 正式训练准备

## 推荐的可靠顺序

不要一开始就启动长训练。以下顺序可以把实际遇到的问题尽早、低成本地隔离出来：

1. 创建隔离 Python 环境，验证 CUDA 与 BF16。
2. 用精确字节数、SHA-256、safetensors 可读性验证每个模型文件。
3. 校验数据集 JSONL：每个目标文件存在，每条 caption 非空。
4. 先构建 latent cache，再构建 text cache，并核对两类 cache 数量。
5. 执行 1-step LoRA 冒烟测试。
6. 记录峰值显存和真实单步时间。
7. 最后才决定分辨率、block swap、checkpoint 间隔和总步数。

1-step 测试必须完整走完：前向、反向、优化器更新、LoRA 保存、Accelerate state 保存。只看到进度条并不代表训练链路已通过。

## 模型文件校验

aria2 的下载结果可能显示 `OK`，但模型仍不可用。旧 `.aria2` 控制文件可能会把截断文件误判为完成；稀疏/预分配文件也可能看起来具有目标大小。

每个模型文件建议同时做三项校验：

```bash
# 1. 精确预期字节数
ls -l /path/to/model.safetensors

# 2. 官方 SHA-256
sha256sum /path/to/model.safetensors

# 3. safetensors 实际可读性
python -c "from safetensors import safe_open; f=safe_open('/path/to/model.safetensors', framework='pt'); print(len(f.keys()))"
```

大文件下载可用的 aria2 参数：

```text
--continue=true --max-connection-per-server=16 --split=16 --min-split-size=20M --file-allocation=none
```

字节数或 SHA 不匹配时，应删除旧控制元数据并重新下载，不能只相信终端摘要。

## Windows 环境隔离：清空 PYTHONPATH

桌面环境中如果同时运行其他 Python 应用，继承的 `PYTHONPATH` 可能将不兼容依赖加载进训练环境。

Windows `.bat` 启动器：

```bat
set "PYTHONPATH="
```

Shell 中：

```bash
env -u PYTHONPATH python ...
```

开始缓存或训练前，应确认导入来自预期虚拟环境。

## 20GB 的 1-step 冒烟测试配置

以下参数足以在 20GB GPU 上验证完整 Ref2VA-Omni 训练链路。这是冒烟测试配置，不是吞吐建议：

```text
--int8_convrot_base
--h3_training_mode ref2va_omni
--network_module networks.lora_minimax_h3
--network_dim 16
--network_alpha 16
--sdpa
--mixed_precision bf16
--gradient_checkpointing
--blocks_to_swap 50
--block_swap_h2d_only
--block_swap_ring_size 1
--block_swap_granularity layer
--optimizer_type AdamW8bit
--learning_rate 1e-4
--save_state
--save_state_on_train_end
--max_train_steps 1
```

它能让 1-step 路径在 20GB 中运行。但不要把它原样套到大显存卡上：50 层 swap 是刻意保守的，会严重牺牲吞吐。

## 48GB 起测配置

针对 `768x1344`、73 帧、batch size 1，不要假设可以无 swap，先从以下参数测起：

```text
--gradient_checkpointing
--blocks_to_swap 24
--block_swap_h2d_only
--block_swap_ring_size 2
```

如果 OOM，逐步提高 swap，例如 32、40；如果能运行，逐步减少 swap 以提升吞吐。我们在 48GB 上用 40 层 swap 获得了稳定运行，但正式训练约为每优化步 123 秒。这是该特定配置的真实测量值，不是通用性能承诺。

每次改动分辨率、帧数、模型、swap、参考模式或 LoRA rank 后，都应重新做 1-step 验证。

## 数据缓存完整性

训练前确认：

- latent cache 数量等于目标样本数；
- text cache 数量等于目标样本数；
- 每对 cache 均可读取；
- cache 的 shape/分辨率与当前 TOML 一致。

正常训练时 trainer 不会自动重建缺失 cache。

## 训练 / 验证划分：必须独立 cache 目录

不要将训练和验证 TOML 指向同一个 cache 目录。H3 bucket 构建过程会枚举目录中的 cache 文件，可能把 JSONL 划分外的样本带入训练。

请使用独立目录，例如：

```text
cache/train_768/  # 只放训练集 latent + text pair
cache/val_768/    # 只放验证集 latent + text pair
```

然后在日志中确认训练样本数与 batch 数严格等于预期值。

完整复现与上游修复建议：#1060。

## INT8 反向兼容性

发布的 pruned INT8 ConvRot Ref2VA checkpoint 在默认 INT8 反向路径上，可能因非 16 对齐的裁剪维度报 cuBLAS unsupported-status 错误。我们暂时使用已有的 transient/BF16 路径作为本地 workaround，并通过完整 1-step LoRA 验证。

详细问题与建议修复：#1059。

在应用任何本地源码改动前，先检查上游 issue 是否已在你的 musubi-tuner revision 中修复。

## 可中断租卡上的长训练

训练应放进 tmux 或其他会话管理器。频繁保存完整 Accelerate state，而不只是 LoRA 权重：

```text
--save_every_n_steps 20
--save_state
--autoresume
```

完整 state 应包括模型、优化器、scheduler、dataloader sampler 与随机状态。长训练前先实际验证一次 resume。

## 监控

建议用 TensorBoard 观察：

- `loss/current`：噪声较大的单 batch 指标；
- `loss/average`：本地累计均值；恢复训练创建新 run 时会重新累计；
- `val/loss`、`val/loss/video`、`val/loss/audio`：判断过拟合最有价值；
- `grad/norm`：检测梯度爆炸或非法梯度。

单个验证点不能构成趋势。小验证集应结合多个 checkpoint，使用同一验证集、固定提示词与固定 seed 对比生成结果。

## 实用停止标准

Flow-matching loss 没有通用的绝对合格值。更可靠的判断：

- train loss 下降，validation loss 下降或稳定：继续；
- train loss 下降，validation loss 在多个 checkpoint 持续上升：停止或选更早 checkpoint；
- NaN/Inf、持续的梯度异常增长、验证失败：先排查再继续花 GPU 费用。

小视频数据集应在多个 checkpoint 以固定提示词采样对比，不要默认最终 step 最好。

## 总结

关键不是说 20GB GPU 很快，而是可以先用它验证完整链路，再租更贵的机器。模型完整性校验、cache 计数、1-step 反向验证、独立验证 cache、频繁保存可恢复状态，能避免大多数昂贵的训练失败。
